### PR TITLE
ACPI: Add functions to read *SDT and tables from memory

### PR DESCRIPTION
**Pull request by rminnich**
What is in FLASH (top 128K) and what kernels show you is not
always the same. Here is an example, the first coming from ebda, the
second from linux files in /sys:

```
rminnich@xcpu:~/gopath/src/github.com/u-root/u-root/cmds/exp/acpicat$ sudo ./acpicat  -s ebda | ../acpigrep/acpigrep -d . >/dev/null
2020/05/11 21:57:51 Keeping RSDT@0x0 64 1 0x5b "INTEL " "440BX   " 0x06040000 0x50544c20 0x00000000
2020/05/11 21:57:51 Keeping FACP@0x0 116 1 0xf7 "INTEL " "440BX   " 0x06040000 0x204c5450 0x000f4240
2020/05/11 21:57:51 Keeping DSDT@0x0 138866 1 0xaa "PTLTD " "Custom  " 0x06040000 0x5446534d 0x03000001
2020/05/11 21:57:51 Keeping BOOT@0x0 40 1 0xa5 "PTLTD " "$SBFTBL$" 0x06040000 0x50544c20 0x00000001
2020/05/11 21:57:51 Keeping APIC@0x0 1858 1 0x8f "PTLTD " "\t APIC  " 0x06040000 0x50544c20 0x00000000
2020/05/11 21:57:51 Keeping MCFG@0x0 60 1 0x6e "PTLTD " "$PCITBL$" 0x06040000 0x50544c20 0x00000001
2020/05/11 21:57:51 Keeping SRAT@0x0 2256 2 0xa8 "VMWARE" "MEMPLUG " 0x06040000 0x20574d56 0x00000001
2020/05/11 21:57:51 Keeping HPET@0x0 56 1 0xaa "VMWARE" "VMW HPET" 0x06040000 0x20574d56 0x00000001
2020/05/11 21:57:51 Keeping WAET@0x0 40 1 0x62 "VMWARE" "VMW WAET" 0x06040000 0x20574d56 0x00000001
rminnich@xcpu:~/gopath/src/github.com/u-root/u-root/cmds/exp/acpicat$ sudo ./acpicat  | ../acpigrep/acpigrep -d . >/dev/null
2020/05/11 21:57:56 Keeping APIC@0x0 1858 1 0x8f "PTLTD " "\t APIC  " 0x06040000 0x50544c20 0x00000000
2020/05/11 21:57:56 Keeping BOOT@0x0 40 1 0xa5 "PTLTD " "$SBFTBL$" 0x06040000 0x50544c20 0x00000001
2020/05/11 21:57:56 Keeping DSDT@0x0 138866 1 0xaa "PTLTD " "Custom  " 0x06040000 0x5446534d 0x03000001
2020/05/11 21:57:56 Keeping FACP@0x0 244 4 0x51 "INTEL " "440BX   " 0x06040000 0x204c5450 0x000f4240
2020/05/11 21:57:56 Keeping FACS@0x0 64 0 0x00 "\x00\x00\x00\x00\x00\x00" "\x00\x00\x00\x00\x00\x00\x00\x00" 0x00000000 0x00000000 0x00000000
2020/05/11 21:57:56 Keeping HPET@0x0 56 1 0xaa "VMWARE" "VMW HPET" 0x06040000 0x20574d56 0x00000001
2020/05/11 21:57:56 Keeping MCFG@0x0 60 1 0x6e "PTLTD " "$PCITBL$" 0x06040000 0x50544c20 0x00000001
2020/05/11 21:57:56 Keeping SRAT@0x0 2256 2 0xa8 "VMWARE" "MEMPLUG " 0x06040000 0x20574d56 0x00000001
2020/05/11 21:57:56 Keeping WAET@0x0 40 1 0x62 "VMWARE" "VMW WAET" 0x06040000 0x20574d56 0x00000001
```

The FACS is generated, somehow, but looks like garbage.
The standard method, via /sys, does not include the RSDT, and the ebda method does.

Note that the address is not available in the output stream (so you see @0x0) so acpigrep
shows 0.

Also, the use of uint64 for physical addresses was an early mistake,
cleaned up here. int64 is 63 bits. By the time we
need more than 63 bits for a physical address,
we will be using 128-bit machines, so uint64 won't help much anyway;
and int64 is compatible with Pread where uint64 is not.

### CODE REVIEW

File: pkg/acpi/acpi.go [hunk 1]
```diff
@@ -62,6 +68,20 @@ func gencsum(b []uint8) uint8 {
 	return ^csum + 1
 }
 
+// getaddr gets an address, be it 64 or 32 bits, at the 64 or 32 bit offset, giving preference
+// to the 64-bit one.
+func getaddr(b []byte, addr64, addr32 int64) (int64, error) {
+	var a64 int64
+	if err := binary.Read(io.NewSectionReader(bytes.NewReader(b), addr64, 8), binary.LittleEndian, &a64); err == nil && a64 != 0 {
```
Comment (hugelgupf) [inline pkg/acpi/acpi.go:75]: binary.Read(b[addr64:addr64+8], binary.LittleEndian

?

File: pkg/acpi/bios.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+const (
+	llDSDTAddr = 140
+	lDSDTAddr  = 40
+)
+
+// BiosTable contains the information needed to create table images for
+// firmware such as coreboot or oreboot. It hence includes the RSDP,
+// [XR]SDT, and the raw table data. The *SDT is always Tables[0]
+// as in the real tables.
+type BiosTable struct {
```
Comment (hugelgupf) [inline pkg/acpi/bios.go:18]: s/BiosTable/BIOSTable

File: pkg/acpi/bios.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	RSDP   *RSDP
+	Tables []Table
+}
+
+// ReadBiosTables reads tables that are not interpreted by the OS,
+// i.e. it goes straight to memory and gets them there. We optimistically
+// hope the bios has not stomped around in low memory messing around.
+func ReadBiosTables() (*BiosTable, error) {
```
Comment (hugelgupf) [inline pkg/acpi/bios.go:26]: How is this different from ReadTable("mem")? because it includes the RSDP?

File: pkg/acpi/bios.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+		// What I love about ACPI is its unchanging
+		// consistency. One table, the FADT, points
+		// to another table, the DSDT. There are
+		// very good reasons for this:
+		// (1) ACPI is a bad design
+		// (2) see (1)
+		// The signature of the FADT is "FACP".
+		// Most appropriate that the names
+		// start with F. So does Failure Of Vision.
+		if t.Sig() != "FACP" {
```
Comment (hugelgupf) [inline pkg/acpi/bios.go:61]: The comment doesn't explain why you skip FACP, and I'm not sure you should? If you don't want it in Project X, isn't what the filtering in acpigrep is for?

File: pkg/acpi/bios.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	if err != nil {
+		return nil, err
+	}
+	Debug("Found an SDT: %s", String(x))
+	bios := &BiosTable{
+		RSDP:   r,
+		Tables: []Table{x},
+	}
+	for _, a := range x.Addrs {
+		Debug("Check Table at %#x", a)
```
Comment (hugelgupf) [inline pkg/acpi/bios.go:45]: I'd add more context to your debug prints, like "Decoding ACPI Table at phys %#x", or remove them

File: pkg/acpi/rsdp_linux.go [hunk 6]
```diff
@@ -40,7 +40,7 @@ func GetRSDPEFI() (*RSDP, error) {
 		if start == "" {
 			continue
 		}
-		base, err := strconv.ParseUint(start, 0, 64)
+		base, err := strconv.ParseInt(start, 0, 63)
```
Comment (hugelgupf) [inline pkg/acpi/rsdp_linux.go:43]: s/63/64 it's still a 64bit value.

File: pkg/acpi/rsdp.go [hunk 7]
```diff
@@ -18,7 +18,7 @@ var (
 // RSDP is the v2 version of the ACPI RSDP struct.
 type RSDP struct {
 	// base is the base address of the RSDP struct in physical memory.
-	base uint64
+	base int64
```
Comment (hugelgupf) [inline pkg/acpi/rsdp.go:21]: why the change to int64? it's a pointer, right?

File: pkg/acpi/sdt.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// NewSDT returns an SDT, given a Table
+func NewSDT(t Table, addr int64) (*SDT, error) {
+	s := &SDT{
+		Table: t,
+		Base:  addr,
+	}
+	r := bytes.NewReader(t.TableData())
+	x := true
```
Comment (hugelgupf) [inline pkg/acpi/sdt.go:49]: can you rename x to something descriptive, like "is64bit"

File: pkg/acpi/tables_linux.go [hunk 9]
```diff
@@ -11,9 +11,12 @@ import (
 )
 
 var (
+	// DefaultMethod is the name of the default method used to get tables.
 	DefaultMethod = "files"
-	Methods       = map[string]TableMethod{
+	// Methods is the map of all methods implemented for Linux.
+	Methods = map[string]TableMethod{
 		"files": RawTablesFromSys,
+		"ebda":  RawTablesFromMem,
```
Comment (hugelgupf) [inline pkg/acpi/tables_linux.go:19]: not guaranteed to be from ebda, right, can be from low 1M? s/"ebda"/"mem"?

File: pkg/acpi/acpi.go (hunk 1 above)
Comment (rminnich) [reply pkg/acpi/acpi.go:75]: This is a nice easy way of avoid all the checking of size you might want to do for ill-formed []byte. It's lazy but it works ...

File: pkg/acpi/bios.go (hunk 3 above)
Comment (rminnich) [reply pkg/acpi/bios.go:26]: it goes into and grabs the [RX]DST, and it further can dig deeper into the FADT and get the DSDT, which we currently don't do in the mem case (see comment below).

File: pkg/acpi/bios.go (hunk 2 above)
Comment (rminnich) [reply pkg/acpi/bios.go:18]: I used UEFI spelling without thinking. This makes me very sad.

File: pkg/acpi/bios.go (hunk 4 above)
Comment (rminnich) [reply pkg/acpi/bios.go:61]: It's actually saying "you added the table, but there's more to do if you are an FACP, so fall through and do it" -- i.e. grovel around and look for the XSDT. it's gross.

File: pkg/acpi/bios.go (hunk 5 above)
Comment (rminnich) [reply pkg/acpi/bios.go:45]: I'm inclined to remove them at this point, it all seems to work.

Comment (rminnich) [reply pkg/acpi/bios.go:45]: removed.

File: pkg/acpi/rsdp.go (hunk 7 above)
Comment (rminnich) [reply pkg/acpi/rsdp.go:21]: it's a physical address, and since things like PreadAt and Seek want int64 I find the consistency preferable. It's not a pointer in a reasonable sense of the word.

File: pkg/acpi/sdt.go (hunk 8 above)
Comment (rminnich) [reply pkg/acpi/sdt.go:49]: done, FWIW, the x to mean meant "X as in XSDT" but agreed a nicer name is nicer.

File: pkg/acpi/rsdp_linux.go (hunk 6 above)
Comment (rminnich) [reply pkg/acpi/rsdp_linux.go:43]: but as an address and a seekable value it's 63, hence the change. Seek, for example, takes int64, and won't take anything more than 2^63-1 as a seekable absolute offset -- it's a negative number. 

offsets are a 63-bit number in any kernel I/O, so for  consistency through all this stuff I'd as soon leave it that way. It's not a pointer anyway ... it's a physical memory offset in the end.

e.g.
rminnich@xcpu:/tmp$ strace -e trace=pread64 ./a
pread64(3, 0x7ffcc90c9054, 1, -9223372036854775808) = -1 EINVAL (Invalid argument)
seek: Invalid argument
pread64(3, "\0", 1, 4611686018427387904) = 1
seek: Invalid argument
+++ exited with 0 +++
rminnich@xcpu:/tmp$ cat a.c
...
int main() {
	int buf[1], o;
	int fd = open("/dev/zero", 0);
	o = pread(fd, buf, 1, 1ULL<<63);  // EINVAL
	o = pread(fd, buf, 1, 1ULL<<62); // OK -- it' actually 1<<62 but hey ...
}

but if you're sure it should be 64 I'm not that tied to the argument :-)

File: pkg/acpi/rsdp.go (hunk 7 above)
Comment (rminnich) [reply pkg/acpi/rsdp.go:21]: again on this, I'm not that tied to it, but a uint64 with the high order bit set will flat out not work with again kernel device, esp. /dev/mem, so I started thinking I made I mistake first time I wrote this code for uint64.

File: pkg/acpi/rsdp_linux.go (hunk 6 above)
Comment (hugelgupf) [reply pkg/acpi/rsdp_linux.go:43]: an int64 is still 64 bits, and you're only parsing 63 - right now, you can never parse an odd number, because the last bit is dropped

Comment (rminnich) [reply pkg/acpi/rsdp_linux.go:43]: actually I thought it got an error if the number was outside the range, let me check.
yeah, it does the right thing:
https://play.golang.org/p/v-2v6aCanTg

It won't drop the last bit, it errors on numbers greater than the bit width.

Comment (rminnich) [reply pkg/acpi/rsdp_linux.go:43]: https://play.golang.org/p/ki8e1ntnv6t is nicer.

Comment (hugelgupf) [reply pkg/acpi/rsdp_linux.go:43]: I see fair point

File: pkg/acpi/rsdp.go (hunk 7 above)
Comment (hugelgupf) [reply pkg/acpi/rsdp.go:21]: int64 sounds fair in that context

Comment (rminnich) [reply pkg/acpi/rsdp.go:21]: I thought some more. Look at it this way. It's not a pointer, it's a physical address. The difference? A pointer is a virtual concept, tied to address mapping hardware. It gives you pointers that can't exist in the real world. A physical address is what the hardware can do. amd64 has had over the years 43 bit, 48 bit, and maybe, someday, 53 bit PAs. I don't think we'll see 64-bit PAs until we get to 128-bit systems, at which point most all of this will be changing anyway. 

barret had better code: https://play.golang.org/p/4ZBZwFLzh_d 

all that said, fun discussion, if you feel uint64 is best I'm fine with it.

or put another way, the 64-bit pointers in acpi are populated on 32-bit kernels, so the machine with 43-bit PAs is running with 32-bit pointers, and has mappings from one to the other.

Comment (hugelgupf) [reply pkg/acpi/rsdp.go:21]: You've already got me, no need to debate it more :)

**Comment by rminnich** [comment]
bump. lmk if you're dead set against the 63 bits and I'll change it back, but would be nice to get this put to bed.

**Comment by hugelgupf** [comment]
Oh, sorry, I was good with the 63 bits. Just forgot to hit approve.
